Paper: PySPH. Prabhu Ramachandran

prabhuramachandran opened the pull request:
The paper is now complete in terms of content.  I will revise it over the course of the week.

jakevdp commented:
Hi! I'll be reviewing your paper. I plan to do inline comments here as I read it, and will do a final overall comment to let you know I'm finished.

jakevdp commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1022 @@
+The examples are written in a way that makes it easy to extend and also
+perform comparisons between schemes.  These features make PySPH well suited
+for reproducible numerical work.  In fact one of our recent papers was written
+such that every figure in the paper is automatically generated using PySPH.
```
👍

jakevdp commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1022 @@
+usually pre-installed.  On OS X, clang will work as will gcc (which can be
+easily installed using brew_). On Windows the Visual C++ Compiler for Python
+will need to be installed.  Detailed instructions for all these are available
+from the `PySPH documentation`_.
```
You might consider putting a recipe for PySPH on [conda-forge](http://conda-forge.github.io/), especially since one of your goals is to lower the barrier-to-entry for people building SPH solvers. If you go that route, then you'd have a cross-platform installation of pySPH, along with its dependencies, using ``conda install pysph --channel conda-forge``

prabhuramachandran replied:
Thanks for the pointer, I will take a look at this later.  There is one issue with this in that on Windows PySPH will not work with mingw and requires MSVC (which is easy to install but I am not sure this is what conda does).  Using mingw is problematic unless Python itself is compiled with mingw.

jakevdp replied:
Conda-forge actually uses MSVC by default, if I recall correctly.

astrofrog replied:
+1 to conda-forge!

prabhuramachandran replied:
Thanks, but that is rather tangential to the paper so I am going to ignore this particular thread of comments for now.

jakevdp commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1022 @@
+  Enter example number you wish to run:
+
+
+Will provide a listing of the examples prompting for a particular one to run.
```
Consider rewording this sentence

prabhuramachandran replied:
Done, I hope it reads better now.

jakevdp commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1022 @@
+
+This provides a very high-level introduction to PySPH in general.  The next
+section discusses some essential software engineering used in the development
+of PySPH.  This is followed by details on the underlying design of PySPH.
```
This road-map would be better located at the beginning of the section

prabhuramachandran replied:
I have not really addressed this as there is a roadmap at the beginning of the section.  This particular overview was rather long so I put the roadmap here again so it reads nicely.  I am happy to make more changes if you feel that it helps.

jakevdp commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1022 @@
+                    dest='fluid', sources=['fluid'],
+                    alpha=self.alpha, beta=0.0, c0=self.co),
+
+                XSPHCorrection(dest='fluid', sources=['fluid']),
```
This line is over-running the column in the PDF

prabhuramachandran replied:
Should be fixed now.

jakevdp commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1022 @@
+exists the code is not recompiled.  Care is taken to look for changes in
+dependencies of this generated source.
+
+As a result of this, our code performs as well as a hand-written FOTRAN code.
```
maybe nitpicking, but does 1.3x qualify as "as well"? This could probably be re-worded. Same with similar language at the beginning of the paper.

prabhuramachandran replied:
Well, it is difficult to perform an exact comparison. I have tried to explain that there is a difference in the implementations.  Even without this a 1.3x factor is certainly within the ballpark.  Given that they use a slightly different algorithm (with a potential for 1.5x speedups) the 1.3 factor is not a significant difference.  Regardless, I have changed the wording a bit.  I have also added some numbers showing excellent scale-up with OpenMP.  Of course a full-fledged performance analysis would be out of scope in this article.

jakevdp commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1022 @@
+  it easy to use PySPH in an automation framework.
+
+Recently, we have used these features to make an entire publication completely
+reproducible.  Every figure produced in the paper (a total of 23 in number)
```
Which paper?

prabhuramachandran replied:
I just submitted it a couple of weeks ago.  I am happy to add a reference but it is not reviewed yet.  I am also afraid that if I do add it, I will exceed the page limits and it looks like with the current changes, it exceeds the 8 page limit.

jakevdp replied:
Fine, but readers are still going to wonder what "the paper" is. I'd either add more to clue the readers in to what you're talking about, or remove this reference to it.

prabhuramachandran replied:
My problem is really that I am running out of space.  If you feel that an additional page is OK, I can add the reference without any problem.  Otherwise I am going to have to make significant edits which is not something I believe I will have time for.  I will try to make some edits today to address your issues and also the page constraint.

jakevdp commented:
Very nice writeup, and it looks like a useful package!

No major critiques, but one minor comment: the paper makes a couple hand-wavy performance claims (e.g. "performs as well as a hand-written FOTRAN code."; "produces good scale-up". I think it would be useful to the reader to quantify these statements, and perhaps include explicit benchmarks.

Other than those comments, I think this is in very good shape.

prabhuramachandran commented:
@jakevdp -- thanks for the review and the kind comments.  I am traveling and will incorporate your suggestions in a few days and respond to the individual comments when I do so.  Thanks again!

prabhuramachandran commented:
@jakevdp -- Thanks again for the review and kind comments!  I have pushed changes to address almost all of the issues.  A full-fledged performance comparison would be a bit out of scope in the current paper.  I have still not addressed issues of MPI scale-up but have added one single piece of evidence about the OpenMP scale up.  The only problem currently is that I think the PDF exceeds the 8 page limit. Advice would be appreciated.  Thanks!

suchow commented:
Here's some advice from Proselint (http://github.com/amperser/Proselint), a Python-based linter for prose:

```
79:23: consistency.spacing Inconsistent spacing after period (1 vs. 2 spaces). Found 15 times elsewhere.
646:50: garner.phrasal_adjectives.ly No hyphen is necessary in phrasal adjectives with an adverb ending in -ly.
1007:1: redundancy.garner Redundancy. Use 'plans' instead of 'Future plans'.
517:70: skunked_terms.misc 'hopefully' is a bit of a skunked term — impossible to use without issue. Find some other way to say it.
274:26: weasel_words.very Substitute 'damn' every time you're inclined to write 'very;' your editor will delete it and the writing will be just as it should be. Found 9 times elsewhere.
```

jakevdp commented:
Quick heads-up: the deadline for review is July 10th, but I'll be out-of-touch after July 6th. If you want my feedback/sign-off on this, it will have to happen before then.

astrofrog commented:
Hi @prabhuramachandran, I am one of the other reviewers for this paper - I'm very sorry for not sending comments before! I will go through now and do my review.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+.. class:: abstract
+
+    Smoothed Particle Hydrodynamics (SPH) is a general purpose technique to
+    numerically compute the solutions to partial differential equations.  The
```
I think you should say which partial differential equations - e.g. *to the partial differential equations for fluid dynamics* or something similar? (it's not a technique to solve any partial differential equation)

prabhuramachandran replied:
Technically you can solve pretty much any PDE although you might not want to do so.  Some problems are better solved with other techniques.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+    Smoothed Particle Hydrodynamics (SPH) is a general purpose technique to
+    numerically compute the solutions to partial differential equations.  The
+    method is grid-free and uses particles to discretize the various
+    properties of interest.  The method is Lagrangian and particles are moved
```
Since this has to be approachable to non-specialists, how about *properties of interest (such as density or temperature)*

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+    to use on multiple platforms, high-performance and support parallel
+    execution.  Users write pure-Python code and HPC code is generated on the
+    fly, compiled, and executed.  PySPH supports OpenMP and MPI for
+    distributed computing.  This is transparent to the user.  PySPH is also
```
The sentence *This is transparent to the user* is a little short - how about *PySPH supports OpenMP and MPI for
distributed computing, in a way that is transparent to the user*

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+----------------------------
+
+SPH (Smoothed Particle Hydrodynamics) is a general purpose technique to
+numerically compute the solutions to partial differential equations.  The
```
As before, you need to mention that this is in the context of fluid dynamics

prabhuramachandran replied:
Not necessary, PySPH can simulate solid mechanics problems too.  But you have a point in that it does not support astro-physical scenarios (yet).

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+The SPH method is relatively easy to implement.  This has resulted in a large
+number of schemes and implementations proposed by various researchers.  It is
+often difficult to reproduce published results due to the variety of
+implementations.  While a few standard packages like (SPHysics
```
Sentences should make sense without the content in brackets, but currently this one reads *standard packages like exist,* - I would suggest instead saying *standard packages exist (such as ...), they are*

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+Our initial implementation was based on Cython :cite:`behnel2010cython` and
+also featured some parallelization using MPI.  This was presented at SciPy
+2010 :cite:`PR-pysph-scipy-2010`.  Unfortunately, this previous version of
+PySPH, proved difficult to use as users were forced to implement most of their
```
Remove comma before 'proved'

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+:cite:`josephine-sph:cpc:2015`, GADGET-2 :cite:`gadget2-springel:mnras:2005`
+etc.)  exist, they are usually tailor-made for particular applications and are
+not general purpose.  They are all implemented in FORTRAN (77 or 90) or C.
+They do not have a convenient Python interface.
```
Maybe combine the last two sentences: *They are all implemented in FORTRAN (77 or 90) or C, and do not have a convenient Python interface.*

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+code in Cython.  This was not a matter of simply writing a few high
+performance functions in Cython.  Our library is object oriented and
+supporting a new SPH formulation would require subclassing one or more classes
+and this would need to be done with Cython.  Our type system also ended up
```
What do you mean by 'Our type system'?

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+becoming more rigid because if we desired performance as well as a general
+purpose code, all the types needed to be pre-defined.  In addition, writing
+all this in Cython meant that users had to manage compilation and linking the
+Cython code during development.  This was not pleasant.
```
The last sentence is a little colloquial for a published proceedings - I would suggest removing it

prabhuramachandran replied:
Totally agree, will fix these and other issues you have kindly pointed out.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+It was felt that we might as well have implemented it all in C++ and exposed a
+Python interface to that.  A traditional compiled language has more developer
+tooling around it.  For example debugging, performance tuning, profiling would
+all be easier if everything were written in C or C++.  Unfortunately, such a
```
Is it really true that a C/C++ code would be easier to debug than a Python code?

prabhuramachandran replied:
Yes, it is in general easier to debug because there is a possibility of controlling everything (you will build everything from scratch pretty much) and you do not have to go across language boundaries.  Cross language debugging is non-trivial, you have to deal with Python, Cython, C/C++ and gdb.  This is vastly easier on Windows with the Python tools for MSVC but that is not what everyone uses.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+script using PySPH can be run with almost no changes using MPI.
+
+PySPH features a reasonable test-suite and we use continuous integration
+servers to test it on Linux and Windows.  Our documentation is hosted on
```
hosted **at**

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+PySPH features a reasonable test-suite and we use continuous integration
+servers to test it on Linux and Windows.  Our documentation is hosted on
+http://pysph.readthedocs.org.  The framework supports several of the standard
+SPH schemes.  A suite of about 30 examples are provided.  These are shipped as
```
The non-expert might not know what a 'scheme' is at this point?

prabhuramachandran replied:
This is referred to in the second paragraph and I have added another sentence there to elaborate this a bit.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+Smoothed Particle Hydrodynamics
+-------------------------------
+
+The SPH method works by approximating the identity:
```
I feel like there are a few sentences of intro missing here - SPH is not just about approximating a delta function, but first of all it should probably be mentioned that it is about discretizing a continuous fluid? Otherwise it's not clear at this point what the delta function refers to physically (this should be clarified)

prabhuramachandran replied:
The equation basically says that a continuous function can be approximated by an integral over nearby function values.  It is exact if one uses a Dirac-delta and can be approximated. 
This is a bit difficult to go into without a larger page limit.  I have already exceeded my 8 page limit and this paper is not an introduction to SPH.  I have cited a few references for this.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+hence provide functionality to:
+
+1. Easily discretize properties into particles.  This is easily done with
+   ``numpy`` arrays representing the property values in Python.
```
Numpy doesn't do the discretization, it allows the representation of the discrete properties - I think this should be clarified.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+Of the above, the NNPS algorithm is usually a well-known algorithm.  For
+incompressible flows where the smoothing radius of the particles, :math:`h`,
+is constant, a simple bin-based linked list implementation is standard.  For
+cases where :math:`h` varies, a tree-based algorithm is typically used.  Users
```
I think you need to explain more for non-experts - I think at this point (or somewhere) you need to mention whether you include things like gravity - in which case all particles influence all others, so searching for nearest neighbors isn't going to be enough.

prabhuramachandran replied:
Good point, I will add a note for this.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+Monaghan :cite:`monaghan-review:2005`.  It is easy to see that equations
+:ref:`eq:sph-continuity` and :ref:`eq:sph-momentum` are ordinary differential
+equations that govern the rate of change of the density and velocity of a
+fluid particle.  In principle, one can integrate these ODEs to obtain the flow
```
ODE hasn't been defined

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+  feature requires mpi4py_ and Zoltan_ to be installed.
+- PySPH provides a built-in 3D viewer for the particle data generated.  The
+  viewer requires Mayavi_ :cite:`it:mayavi:cise:gael2011` To be installed.
+- PySPH is also free and currently hosted at http://pysph.bitbucket.org
```
'free' is an ambiguous terms :) Maybe just say 'open-source'?

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+mpi4py_ is also needed in this case.  Unfortunately, MPI is not tested on
+Windows by us currently.
+
+PySPH also provides an optional 3D viewer and this depends on Mayavi_.
```
I would avoid one-sentence paragraphs if possible

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+PySPH also provides an optional 3D viewer and this depends on Mayavi_.
+
+In summary, PySPH is easy to install if one has a C++ compiler installed.
+MPI support is a little involved due to the requirement to install Zoltan_.
```
What is Zoltan and why can't one just use mpi4py?

prabhuramachandran replied:
Zoltan does a lot more as it manages the distribution and load-balancing of particles.  I can add a line for this.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+~~~~~~~~~~~~~~~~~~~
+
+PySPH is tested to work with Python-2.6.x to 2.7.x and also with Python
+3.4/3.5.  PySPH is a typical Python package and can be installed fairly easily
```
Just out of curiosity, does Mayavi work in Python 3 yet?

prabhuramachandran replied:
Yes.  4.4.4 added experimental Python 3 support.  The problem was not mayavi really, it was relatively easy to have mayavi ported to Python3, VTK was a bigger issue and with version 7.0 it does support Python3.  This needs more testing.  The bleeding edge does with with Python 3 and I should push a release sometime.  Hopefully before scipy.

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+rescan the directory to check for any new files.  This makes it convenient to
+visualize the results from a running simulation.  The "Connection" tab can be
+used when the visualization is in "Live mode" when it can connect to a running
+simulation and view the data live.  While this is very cool in principle, it
```
I would avoid the word 'cool' since it is very colloquial - even though it really is cool :)

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+modern open source implementations follow.  For example:
+
+- Our sources are hosted on bitbucket (http://pysph.bitbucket.org).  We are
+  thinking of shifting to github because github has much better integration
```
github -> GitHub (here and elsewhere, when used in the main text)

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+  thinking of shifting to github because github has much better integration
+  with continuous integration services and this is a rather frustrating pain
+  point with bitbucket.
+- We use pull-requests to review all new features and bug fixes.  At this
```
I don't think pull requests should be hyphenated?

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+  builds.
+- Our documentation is generated using Sphinx and hosted online on
+  http://pysph.readthedocs.io.
+- Releases are pushed to pypi.
```
pypi -> PyPI

Maybe spell it out: *Releases are pushed to the Python Package Index (PyPI)*

astrofrog commented on papers/prabhu_ramachandran_pysph/prabhu_ramachandran_pysph.rst:
```diff
@@ -0,0 +1,1044 @@
+etc.  The arrays are all one dimensional.  The indices that are outside the
+circle are identified between lines 11 and 14 and these are removed in
+line 20.  This could have also been done with pure NumPy indexing.  In Line 17
+the particle array instance is created and is called ``"fluid"``.  Line 22
```
Remove ``"``?

astrofrog commented:
Overall, this looks good and PySPH seems like a very interesting package.

One thing I feel is missing is a little more discussion in the introduction sections about what kind of physics are supported (the intro to SPH jumps right in to the mathematical details). For instance, for Astrophysical applications, one needs to take into account gravity, which is usually the bottleneck for performance. Can PySPH be used for this?  If so, does the user have to implement the gravity tree-code themselves? Either way, can you clarify which fields PySPH of science would be applicable to?

prabhuramachandran commented:
@astrofrog -- thanks for the detailed comments!  I will take a look and incorporate most of your comments and try to fix this and other issues by this weekend.  I am a bit worried about the page limit though as I have already exceeded 8 pages.

prabhuramachandran commented:
@suchow -- thanks for this and the link to proselint, I will check it out.

prabhuramachandran commented:
@jakevdp -- I've made the changes you had requested.  I added a reference to the paper I had mentioned (unfortunately this is just under review but readers can find it if needed when it does get accepted/published).  I have also edited some of the pre-formatted code blocks for the `pysph run` example and the openmp comparison to just barely fit the paper into 8 pages.  I will next address @astrofrog's comments.  I hope this should address all your concerns. Thanks!

prabhuramachandran commented:
@astrofrog -- I think I have made the necessary changes suggested by you.  Please let me now if this is now OK.  I have also made some changes so the manuscript just fits in 8 pages.

jakevdp commented:
Looks good to me. Thanks @prabhuramachandran!

sarostru commented:
Due to the complication of the proceedings deadline lining up with everyone's travel schedule to attend the conference we've decided to allow authors to spend an additional 2 days, that is until July 12th to complete any major changes to the submission.  This may impact your enjoyment of the conference, so we hope not too many of you need to take the time.  For reviewers, we ask that you read over this final document before the conference is over and post your decision here in the comments on or before July 17th.

Thanks and have a great Scipy!
Your proceedings co-chairs Scott Rostrup (@sarostru) and Sebastian Benthall (@sbenthall)

astrofrog commented:
@prabhuramachandran - this looks good to me.

I accept this paper to be merged

sarostru commented:
@prabhuramachandran Looks like the reviewers are ready to merge so I've added the ready tag.  Next up is preparing for the final camera-ready version, so please go ahead and carefully reread, looking for any formatting or language errors, and @sbenthall and I will help with updating the proceedings build from draft to final.
